### ISSUE TITLE
Secure password textbox (vgui)

### ISSUE THREAD

**Issue by CaptainPRICE** [Facepunch/garrysmod-requests#1367]
Think of it as `DTextEntry`, but specialized for entering sensitive user info (i.e. password).

A couple of implementation details to note:
- There should be no way of "getting/setting user entered input" at Lua-level. A new class/type is required to achieve this functionality (a.k.a. `SecureString`), which would work as a userdata at Lua-level, and only C-level would be able to read/use its contents.
- Copy/Paste (shortcuts) should be restricted/disabled (including Select All). Additionally, right-click context menu should be disabled too.

A couple of features to note:
- Do allow to change the `PasswordChar` (a character displayed in box, by default `●`).
- Do allow to set `MaxLength` (maximum length of user input).

**Comment by thegrb93** [comment]
>There should be no way of "getting/setting user entered input" at Lua-level

sooooooooo what's it used for then?

**Comment by CaptainPRICE** [comment]
> sooooooooo what's it used for then?

Used at C level, if a value is of type `SecureString` it would be able to read its contents.
Basically, **any** GMod Lua function, which takes a `string` should work with this design.
When C level recognizes a value is `SecureString`, it would (lock and) *unsecure* the userdata to read its value (and unlock it again).
As an example usage, take Rest API ([`HTTP`](https://wiki.garrysmod.com/page/Structures/HTTPRequest)) for example, where you need to send user credentials using HTTP request.
Because, right now it is all plain text and visible to GMod Lua, this is kind of hurting security (IMO).

**Comment by meepen** [comment]
Even if this were to exist, what's stopping someone from mimicking the
design of this Lua interactive textbox?

**Comment by CaptainPRICE** [comment]
> what's stopping someone from mimicking the
> design of this Lua interactive textbox?

Such aspect is certainly offtopic..
That's like you're asking "what's stopping someone from hacking the server".

**Comment by thegrb93** [comment]
>Such aspect is certainly offtopic..

The whole point of this is supposedly to stop server owners from sniffing http credentials. He's proposing they simply mitigate it by mimicking the UI. Thus making it completely useless of an idea.

**Comment by thegrb93** [comment]
tl;dr this would bring false sense of security and is a very dangerous and bad idea.

**Comment by CaptainPRICE** [comment]
> The whole point of this is supposedly to stop server owners from sniffing http credentials

Yeah, in the case of HTTP (example usage I had put above).

>  (you've still not explained the purpose).

The purpose is obvious by now.. isn't it?

> He's proposing they simply mitigate it by mimicking the UI.

I get that. Didn't you read my response comment?

> Thus making it completely useless of an idea.

Ok, could you explain why would server want to prompt a user and intentionally try to steal *secure* info?
I'm kind of repeating myself, so here: "Because, right now it is all plain text and visible to GMod Lua"
In that sense, it is already visible to the server (owners).

Ask yourself the following: Why would server want to prompt a user with `DTextEntry`, to let user enter their password? (Which is plain textbox, current design - could call it the *infected* textbox too.)

(Also, if a client is interested/paranoid, it is possible to know whether particular vgui control is authentic/real, in this case "Secure textentry", by examining piece of code at runtime due all Lua being run on clientside is visible to each client.)

> this would give a false sense of security and is a very dangerous and bad idea.

No.
First of all, client & server are unable to know the contents of userdata (even if you dump game memory on client, you should **not** be able to find the entered input).
Because of secure implementation, using "Secure textentry" is already way **better** (than plain textentry), and certainly more secure than prompting a user with plain DTextEntry.
Thus DTextEntry should not be used in such cases, if you're asking a client to enter their password. So, if you don't get it, this means it would be very easy to figure whether vgui is an *infected* DTextEntry or real "Secure textentry".
I said it, "mimic UI" is offtopic here. If server is trying to mimic and fool "Secure textentry" on a client. A server could have just prompted a user with `DTextEntry`.
"Mimic UI" argument is just stupid hence.

Think again. Still don't get the purpose?

**Comment by willox** [comment]
What would code using this look like?

**Comment by thegrb93** [comment]
>Ok, could you explain why would server want to prompt a user and intentionally try to steal secure info?

To get secure info? No shit?

**Comment by thegrb93** [comment]
Secure info shouldn't be entered in-game for any reason. Nothing is safe, not even html panels.

**Comment by CaptainPRICE** [comment]
> Nothing is safe

You clearly don't know how `SecureString` even works in the first place. **This whole thing is based and depends on that class**.
You are thinking that `SecureString` is just a normal (raw) `string`. And there's your problem.

> not even html panels.

Why.. did you brought html panels.. for no good reason?
Yes, html panels are obviously not safe, stop going offtopic. Same goes for every other vgui component, because control's text/html/contents can be directly read from client's memory.
It seems like you still don't understand a single security aspect of why such a thing should be implemented.

> Secure info shouldn't be entered in-game for any reason.

🤦‍♂

**Comment by willox** [comment]
> You clearly don't know how SecureString

I think they don't work. You can open another issue if you actually come up with some sort of specification.
